### PR TITLE
Change LocalStack of batches from global to client-local in storage

### PR DESCRIPTION
**Pull request by dhermes**
Has #989 as diffbase

### CODE REVIEW

**Comment by tseaver** [comment]
@dhermes ping me after #989 lands (a rebase would help review).

**Comment by dhermes** [comment]
Will do

**Comment by dhermes** [comment]
@tseaver Rebased

File: gcloud/storage/blob.py [hunk 1]
```diff
@@ -291,7 +291,7 @@ def download_to_file(self, file_obj, client=None):
             headers['Range'] = 'bytes=0-%d' % (self.chunk_size - 1,)
         request = http_wrapper.Request(download_url, 'GET', headers)
 
-        download.InitializeDownload(request, client.connection.http)
+        download.InitializeDownload(request, client._connection.http)
```
Comment (tseaver) [inline gcloud/storage/blob.py]: Copy the comment on the reach-around here.

File: gcloud/storage/test_batch.py [hunk 2]
```diff
@@ -84,6 +84,18 @@ def test_ctor_w_explicit_connection(self):
         self.assertEqual(len(batch._requests), 0)
         self.assertEqual(len(batch._target_objects), 0)
 
+    def test_current(self):
+        from gcloud.storage.client import Client
+        project = 'PROJECT'
+        credentials = _Credentials()
+        client = Client(project=project, credentials=credentials)
```
Comment (tseaver) [inline gcloud/storage/test_batch.py]: Add a test for `None` when empty?

File: gcloud/storage/test_batch.py [hunk 3]
```diff
@@ -354,39 +366,32 @@ def test_finish_nonempty_non_multipart_response(self):
...
         URL = 'http://example.com/api'
         expected = _Response()
         expected['content-type'] = 'multipart/mixed; boundary="DEADBEEF="'
         http = _HTTP((expected, _THREE_PART_MIME_RESPONSE))
         connection = _Connection(http=http)
-        client = _Client(connection)
+        project = 'PROJECT'
+        credentials = _Credentials()
+        client = Client(project=project, credentials=credentials)
+        client._connection = connection
```
Comment (tseaver) [inline gcloud/storage/test_batch.py]: Using the "real" `Client` but stubbing out its `_connection` attribute seems double-wrong.  FWIW, `datastore.test_batch` just mocks out the whole thing.

File: system_tests/storage.py [hunk 4]
```diff
@@ -51,15 +51,9 @@ def setUp(self):
...
-            # Stop-gap measure to support batches during transation to
-            # to clients from implicit behavior.
-            batch_client = storage.Client(
-                project=CLIENT.project,
-                credentials=CLIENT.connection.credentials)
-            batch_client.connection = batch
+        with storage.Batch(CLIENT):
             for bucket_name in self.case_buckets_to_delete:
-                storage.Bucket(batch_client, name=bucket_name).delete()
+                storage.Bucket(CLIENT, name=bucket_name).delete()
```
Comment (tseaver) [inline system_tests/storage.py]: Yay!

File: gcloud/storage/test_batch.py (hunk 3 above)
Comment (dhermes) [inline gcloud/storage/test_batch.py]: I'm not sure what's double wrong, but the main goal is to avoid the HTTP request. You did make me realize I could just do

```python
client._connection._http = http
```

instead of making a custom connection.

**Comment by dhermes** [comment]
@tseaver I addressed 2 of the 3 comments and responded to the third (slightly changed a bit).

**Comment by tseaver** [comment]
I can let the "real client" bit go, I guess.  LGTM

**Comment by dhermes** [comment]
Thanks!
